File: binv10/first_problem.rb
require 'json'

users = [
  { "name": "Alice", "email": "alice@gmail.com" },
  { "name": "Bob", "email": "bob@yahoo.com" },
  { "name": "Carol", "email": "carol@gmail.com" }
]

groupedUsers = users.group_by { |user| user[:email].split('@')[1] }

filteredUsers = groupedUsers.each_with_object({}) do |(k, v), h|
  h[k] = v.map { |a| a[:name] }
end

puts JSON.pretty_generate({ data: filteredUsers })

File: binv10/first_problem.js
const users = [
  { name: "Alice", email: "alice@gmail.com" },
  { name: "Bob", email: "bob@yahoo.com" },
  { name: "Carol", email: "carol@gmail.com" },
];

for (const user of users) {
  const domain = user?.email.split("@")[1];
  if (!groupedUsersName[domain]) {
    console.log(domain);
    groupedUsersName[domain] = [];
  }

  groupedUsersName[domain].push(user.name);
}

console.log(JSON.stringify({ data: groupedUsersName }, null, 2));
